portfolio/src/main/java/com/google/sps/servlets/AuthStatusServlet.java

```diff
@@ -0,0 +1,62 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.auto.value.AutoValue;
+import com.google.gson.Gson;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that handles authentication: provides login and logout URLs for users
+ * and checks login status of the user
+ */
+@WebServlet("/auth-status")
+public class AuthStatusServlet extends HttpServlet {
+
+  private static final String AUTH_LOGIN_LOGOUT_REDIRECT_URI = "/";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    UserAuthStatus statusData;
+    if (userService.isUserLoggedIn()) {
+      String logoutUrl = userService.createLoginURL(AUTH_LOGIN_LOGOUT_REDIRECT_URI);
+      String userEmail = userService.getCurrentUser().getEmail();
+      statusData = UserAuthStatus.create(true, userEmail, "", logoutUrl);
+    } else {
+      String loginUrl = userService.createLoginURL(AUTH_LOGIN_LOGOUT_REDIRECT_URI);
+      statusData = UserAuthStatus.create(false, "", loginUrl, "");
```

Comment:
When the number of parameters increases, the create() function starts to look confusing. In this case, consider to use the builder version of the autovalue. The builder will change this code to
```
UserAuthStatus.builder()
   .setIsLoggedIn(...)
   .setLoginUrl(...)
   .setLogoutUrl(...)
   .build()
```
builder pattern also support default value, so you don't have to pass empty string here.

Btw, compiled protobuf java code also uses the builder pattern.

Reply:
Makes sense. This would definitely be useful for value classes with a ton of different optional/nullable fields.

However, I just tried to implement this and it type checks and compiles. However, calling `/auth-status` now results in a StackOverflowError on the server side when it calls `AutoValue_UserAuthStatus.builder()`. I'll try debugging this tomorrow; it might be because of my use of AutoValue builders in an already nested class.

Reply:
I think you need to call `new AutoValue_UserAuthStatus.Builder()`, which is an auto generated class. Also, by default you are required to set all values in the builder, but you can preset the value in your `builder()` method

Reply:
oh man I completely forgot to type new and was wondering why the `.Builder()` method didn't exist. Thanks for spotting my typo!!